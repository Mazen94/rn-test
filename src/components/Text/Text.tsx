import React from 'react';
import { Text as RNText, TextStyle } from 'react-native';

import PropTypes from 'prop-types';

import Colors from '_utils/colors';
import { fonts, fontsScales } from '_utils/fontManager';

type PropsType = {
  size: 's' | 'default' | 'm' | 'l' | 'xl';
  weight: 'regular' | 'bold' | 'italic' | 'light' | 'medium';
  color: string;
  align: 'left' | 'right' | 'center';
  underline: boolean;
  useDefaultLineHeight: boolean;
  style?: TextStyle;
  children: any;
};

const Text = (props: PropsType) => {
  const {
    size,
    weight,
    color,
    align,
    underline,
    useDefaultLineHeight,
    style,
    children,
    ...otherProps
  } = props;

  // Build text style from props
  let sizeAttrs;
  const scale = fontsScales[size];
  if (scale) {
    const { fontSize, lineHeight } = scale;
    sizeAttrs = useDefaultLineHeight ? { fontSize } : { fontSize, lineHeight };
  }

  const customStyle: TextStyle = {
    ...(fonts[weight] || {}),
    ...(sizeAttrs || {}),
    color,
    textAlign: align,
  };
  if (underline) customStyle.textDecorationLine = 'underline';

  // Merge with existing style
  let finalStyle;
  if (Array.isArray(style)) {
    finalStyle = [customStyle, ...style];
  } else {
    finalStyle = {
      ...(customStyle || {}),
      ...style,
    };
  }

  return (
    <RNText {...otherProps} style={finalStyle}>
      {children}
    </RNText>
  );
};

Text.defaultProps = {
  size: 'default',
  weight: 'regular',
  color: Colors.TEXT,
  align: 'left',
  underline: false,
  useDefaultLineHeight: false,
};

Text.propTypes = {
  size: PropTypes.oneOf(['s', 'default', 'm', 'l', 'xl']).isRequired,
  weight: PropTypes.oneOf(['regular', 'bold', 'italic', 'light', 'medium'])
    .isRequired,
  color: PropTypes.string.isRequired,
  align: PropTypes.oneOf(['left', 'right', 'center']).isRequired,
  underline: PropTypes.bool.isRequired,
  useDefaultLineHeight: PropTypes.bool.isRequired,
};

export default React.memo(Text);
