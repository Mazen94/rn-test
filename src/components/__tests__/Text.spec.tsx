import React from 'react';
import { Text as RNText } from 'react-native';
import renderer from 'react-test-renderer';

import Text from '../Text/Text';

describe('[Component] Text', () => {
  it('should render a text', () => {
    const tree = renderer.create(<Text>Hello</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with existing style prop (object)', () => {
    const tree = renderer
      .create(<Text style={{ color: 'red', marginTop: 10 }}>Hello</Text>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with native Text children', () => {
    const tree = renderer
      .create(
        <Text>
          <RNText>Hello</RNText> <RNText>world!</RNText>
        </Text>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with a different size', () => {
    const tree = renderer.create(<Text size="m">Hello</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with a different weight', () => {
    const tree = renderer.create(<Text weight="bold">Hello</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with a different color', () => {
    const tree = renderer.create(<Text color="red">Hello</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with a different alignment', () => {
    const tree = renderer.create(<Text align="center">Hello</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an underline text', () => {
    const tree = renderer.create(<Text underline>Hello</Text>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a text with default line height', () => {
    const tree = renderer
      .create(<Text useDefaultLineHeight>Hello</Text>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
