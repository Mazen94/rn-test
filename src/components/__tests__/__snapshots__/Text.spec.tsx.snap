// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Text should render a text 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "textAlign": "left",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with a different alignment 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "textAlign": "center",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with a different color 1`] = `
<Text
  style={
    Object {
      "color": "red",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "textAlign": "left",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with a different size 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-regular",
      "fontSize": 16,
      "fontWeight": "400",
      "lineHeight": 17.6,
      "textAlign": "left",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with a different weight 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-Bold",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "textAlign": "left",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with default line height 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "textAlign": "left",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with existing style prop (object) 1`] = `
<Text
  style={
    Object {
      "color": "red",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "marginTop": 10,
      "textAlign": "left",
    }
  }
>
  Hello
</Text>
`;

exports[`[Component] Text should render a text with native Text children 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "textAlign": "left",
    }
  }
>
  <Text>
    Hello
  </Text>
   
  <Text>
    world!
  </Text>
</Text>
`;

exports[`[Component] Text should render an underline text 1`] = `
<Text
  style={
    Object {
      "color": "rgb(255, 24, 55)",
      "fontFamily": "Roboto-regular",
      "fontSize": 14,
      "fontWeight": "400",
      "lineHeight": 15.400000000000002,
      "textAlign": "left",
      "textDecorationLine": "underline",
    }
  }
>
  Hello
</Text>
`;
